### DISCUSSION TITLE
Documentation clarification: Console runner vs MSBuild runner ?

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#2702] [Question] [answered]
👋🏻 G'Day awesome xUnit team! 

In the [official xUnit docs, specifically about: Running Tests in Parallel - Parallelism in Runners](https://xunit.net/docs/running-tests-in-parallel#parallelism-in-runners) it mentions two types of runners:

- Console
- MSBuild

Can these two please be explained what they are / where they are / etc?

for example, is the "Console runner" literally `dotnet test` ?

And where is this MSBuild runner?

Finally, are these linked to the Visual Studio Test Runner? Is that a GUI wrapper over the console or msbuild runners?

If none of these are related to the `dotnet test` command, how does that fit into all of this, with respect to parallelism and define the parallelism.

cheers!

**Comment by bartelink** [comment, marked as answer]
TL;DR, "The Console Runner" might best be termed "The independent standalone xunit specific test host" if it was to be built today. Originally, in V1, there was a GUI runner, a Console Runner, and an MSBuild runner

----

The console runner is `xunit.console.exe`, which is shipped in a nuget. Before NuGet, it was shipped as an exe - its a free standing runner that requires only .NET - you point it at an assembly containing .NET test code and it discovers and/or runs.

The msbuild runner is a plugin for MSBuild - i.e. a DLL that implements an interface so that you can invoke tests with a specific command (as opposed to doing `<Exec Command="xunit.console.exe"` etc

There is an internal shared lib that is shared between both of these.

---- 

.NET Core added  a top level `dotnet test` tool, which is a generic test running front end that can host anything that adheres to its interfaces.

If you reference `xunit.runner.visualstudio`, you pull in the packages to emit the plugins that `dotnet test` and/or msbuild use to discover and/or trigger tests.

Each of these mechanisms have impls for NUnit, xUnit and MSTest

In general, xunit manages the parallelism options -you're tunneling settings through dotnet test into xunit's test execution engine.

**Reply by bradwilson** [reply]
The MSBuild runner is akin to the Console runner in that it is an xUnit.net specific test host. The documentation page here describes how to use it: https://xunit.net/docs/running-tests-in-msbuild

Note that it only supports .NET Framework test projects at this time, due to issues with dependency resolution under .NET/.NET Core. For v3, our xUnit.net specific test hosts (Console and MSBuild) will add support for running .NET/.NET Core in addition to .NET Framework tests.

**Reply by PureKrome** [reply]
At the end of the day, what does `dotnet test` _really_ end up using? 

> `dotnet test` tool, which is a generic test running front end that can host anything that adheres to its interfaces.

Okay, so this means that this runner is just a facade over _something_ .

> If you reference xunit.runner.visualstudio, you pull in the packages to emit the plugins that dotnet test and/or msbuild use to discover and/or trigger tests.

Okay, so then this means that `dotnet test` will use the `xunit.runner.visualstudio` ... which uses `xunit.console.exe` ?

**Reply by bartelink** [reply]
>Okay, so this means that this runner is just a facade over something .

Yes, it does not directly run tests - iit has protocols for asking a test FW what tests exist, and then for running them and grabbing the outcoms.

> Okay, so then this means that `dotnet test` will use the `xunit.runner.visualstudio` ... which uses `xunit.console.exe` ?

No, `dotnet test` and `xunit.console.exe` are both top level hosts - one is part of .NET and can run any test FW, the other is xunit specific.

There is some shared logic in a DLL - maybe xunit.runner.dll ? And some more in xunit.dll itself

**Reply by bradwilson** [reply]
_This information is for xUnit.net v2. Lots of things are changing for v3._

The layering here can be a bit confusing.

At the very bottom layer of our runner system, we have `xunit.runner.utility.dll`, which is a library that anybody can use to run xUnit.net tests. It understands and hides the differences between, for example, v1 vs. v2 test assemblies. We provide this with both backward-major and forward-minor compatibility (meaning, if you write a runner against the 2.0 version of `xunit.runner.utility.dll`, it would not only be able to run everything written against v1.x, as well as v2.0, but any future v2.x release as well). We will continue this with v3. This is why the libraries you use for **_writing tests_** -- `xunit.core.dll` and `xunit.assert.dll`, most commonly -- are isolated from the libraries you use for _**running tests**_, to prevent any strong version coupling. Runners generally don't know and don't care what version of xUnit.net your tests were written against.

The next layer up, we offer several runners and/or runner plugins. This includes, but is not limited to:

- Purpose-built runners
   - Console runner (`xunit.console.exe`). This currently only supports .NET Framework.
   - MSBuild task (`xunit.runner.msbuild.dll`). This currently only supports .NET Framework.
   - Devices runner (`xunit.runner.devices`). This currently supports Windows 10+ UWP, Xamarin MonoAndroid, and Xamarin iOS Unified. This previously supported Windows 8 Silverlight, Windows 8.1 UAP, and Xamarin MonoTouch; these have all since been deprecated.
- Plugins to generic runners
   - Our TestDriven.net plugin (`xunit.runner.tdnet.dll`). This currently only supports .NET Framework.
   - Our VSTest plugin (`xunit.runner.visualstudio.dll`). This currently supports .NET Framework, .NET/.NET Core, and Windows 10+ UWP. This previously supported Windows 8 Silverlight and Windows 8.1 UAP, both of which have since been deprecated.

Our console runner stands alone and doesn't plug into anything.

Our MSBuild task requires MSBuild to run, but it exposes itself just as an MSBuild task that you can invoke. MSBuild does not contain any unit testing specific logic. As such, the `<xunit>` MSBuild task offers much the same functionality as our console runner, via task properties (rather than command line switches like you'd use in our console runner).

Our Devices runner helps you create an application that runs on your target device to allow you to run your unit tests. It provides a bare-bones UI for this purpose.

Our TestDriven.net plugin supports their plugin model (defined in `TestDriven.Framework.dll`). They support direct testing gestures via mouse and keyboard in the editor of Visual Studio, with output to the VS output window. They do not have any dedicated UI.

Our VSTest plugin supports their plugin model (defined in `Microsoft.TestPlatform.ObjectModel.dll`). On top of their plugin system, they expose several ways to run tests:

- The Test Explorer window in Visual Studio
- The Test panel in Visual Studio Code (via the C# VS Code extension)
- `dotnet test`
- `vstest.console.exe`

Both the TDnet and VSTest plugin models are framework agnostic; anybody who wants to plug in there must provide an adapter library. Even MSTest is written via the same plugin model (provided by `Microsoft.VisualStudio.TestPlatform.Extensions.VSTestIntegration.dll`). The UI here, in particular the command line switches and available customization, differ dramatically from our purpose-built runners, because these platforms are framework agnostic and offer mostly lowest-common-denominator features.

**Reply by PureKrome** [reply]
thank you Brad for the lovely answer!
